🏃 Move MHC remediations to MachineSet controller

benmoss opened the pull request:
**What this PR does / why we need it**:
- Changes the MachineHealthCheck controller to only annotate unhealthy machines with the new annotation
- Changes the MachineSet controller to now remediate machines with this annotation by deleting them

Follow up from #2920

JoelSpeed commented on controllers/machinehealthcheck_targets_test.go:
```diff
@@ -337,68 +331,23 @@ func TestHealthCheckTargets(t *testing.T) {
-			expectErr:     false,
-			expectDeleted: false,
-			expectEvents:  []string{},
+			name:            "it annotates the machine with the unhealthy annotation",
```
Would be good to add another test case here that has existing annotations on the machine and check that any existing annotations are not dropped as part of this annotation process

JoelSpeed commented on controllers/machinehealthcheck_targets_test.go:
```diff
@@ -409,232 +358,28 @@ func TestRemediate(t *testing.T) {
-			if tc.expectDeleted {
-				gs.Expect(errors.IsNotFound(err)).To(BeTrue())
+			if tc.expectAnnotated {
+				gs.Expect(machine.GetAnnotations()).To(HaveKey(clusterv1.MachineUnhealthy))
```
```suggestion
				gs.Expect(machine.GetAnnotations()).To(HaveKeyWithValue(clusterv1.MachineUnhealthy, ""))
```

JoelSpeed commented on controllers/machineset_controller.go:
```diff
@@ -59,6 +59,11 @@ var (
 )
 
+//go:generate counterfeiter . machineHealthCheck
+type machineHealthCheck interface {
```
This isn't actually healthy checking, perhaps it should be `machineRemediator` instead?

JoelSpeed commented on controllers/machineset_controller.go:
```diff
@@ -203,6 +209,12 @@ func (r *MachineSetReconciler) reconcile(ctx context.Context, cluster *clusterv1
+	if remediated, err := r.MachineHealthCheck.Remediate(ctx, allMachines.Items); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to remediate machines")
+	} else if remediated {
+		return ctrl.Result{Requeue: true}, nil
```
Probably worth a comment explaining why we are requeueing here

detiber replied:
Is immediate requeueing the expected goal here, or should this use RequeueAfter?

It seems like the next requeue will likely return the same list of machines with the `r.Client.List` call on line 205 that we would have if we just continued on here (assuming we are reflecting changes made during remediation to the resources to allMachines).

benmoss replied:
Yeah, I didn't know that `Requeue: true` was immediate, I thought it just meant after some normal interval we requeue.

detiber replied:
It depends...  I believe it is subject to the exponential backoff, but assuming an initial reconciliation for the resource in the queue, the first requeue would effectively be immediate.

vincepri replied:
Do we actually need to force-requeue vs just return? Wouldn't we get a notification for the deletion timestamp being set on the Machine in the MS anyways?

vincepri commented on Makefile:
```diff
@@ -54,6 +54,7 @@ GO_APIDIFF := $(TOOLS_DIR)/$(GO_APIDIFF_BIN)
 # Binaries.
 # Need to use abspath so we can invoke these from subdirectories
 KUSTOMIZE := $(abspath $(TOOLS_BIN_DIR)/kustomize)
+COUNTERFEITER := $(abspath $(TOOLS_BIN_DIR)/counterfeiter)
```
I'd rather avoid using mocks to test controller behavior. Instead, we can test the full controller behavior adding an integration test in the current MachineSet ginkgo tests

benmoss replied:
The idea here is that the remediation is not the controller's behavior, it's the responsibility of the remediator that the controller depends on but does not need to understand. It's a contract that you call `Remediate` and give it machines and it will tell you whether any remediation happened or not.

A unit test is for testing one unit at a time, so here we can assert both sides of the contract but not the interdependency. I agree we should have some tests that do test these things together, I just am arguing they're not unit tests. What I am trying to do here is break up code that is in my opinion way too coupled. The controllers are the entrypoint to the system but that does not mean that we can't have units of code that are not just subcomponents of the controller.

vincepri replied:
> It's a contract that you call Remediate and give it machines and it will tell you whether any remediation happened or not.

The contract, from an API point of view, states that the only expectation for the owning-controller is to delete the Machine. Code shouldn't be part of any contract, one example is controllers not written in Go, or not using Cluster API as a library. They should be able to function by looking at the data and make decisions, ultimately satisfy the API contract expectations.

While it's fine to have unit tests in the exposed remediation functions, I'd like to avoid to test mocked tests when we have the ability to run an integration test which is going to be much closer to real world usage.

How we structure code is a different thing, the remediation code can be tested separately.

ncdc replied:
I am strongly opposed to mocks in the majority of cases. Speaking from experience, mocking behavior makes your tests extremely brittle. If you change the internals of your implementation, this frequently requires you to adjust the mocks' expectations. This may seem reasonable and easy if it's a couple of days after the initial implementation is committed to the repo, if you (or worse, someone else) come back weeks later, it's often very difficult to page things back in to memory. This is made worse if you have complicated test logic (if xyz set, expect mock to do this; else don't expect it) and tons of tests.

benmoss replied:
This is the result of poorly written code and tests and not the fault of the mocks. I'm not interested in fighting this fight anymore so I'll remove them.

vincepri commented on api/v1alpha3/machine_types.go:
```diff
@@ -29,6 +29,9 @@ const (
 	// MachineControlPlaneLabelName is the label set on machines or related objects that are part of a control plane.
 	MachineControlPlaneLabelName = "cluster.x-k8s.io/control-plane"
 
+	// MachineUnhealthy is set on machines determined unhealthy by the Machine Health Check controller.
```
The comment should be explicit about the contract, e.g. expect a machine to be deleted by the owning controller, etc.

benmoss replied:
Added another line, is this what you had in mind?

JoelSpeed replied:
Looks like the extra comment line may have been dropped?

benmoss replied:
Indeed, somehow messed something up in rebasing it seems

benmoss commented on controllers/machinehealthcheck_targets.go:
```diff
@@ -274,86 +262,24 @@ func minDuration(durations []time.Duration) time.Duration {
-		r.Eventf(
-			t.Machine,
-			corev1.EventTypeWarning,
-			EventMachineDeletionFailed,
-			"Machine %v remediation failed: unable to delete Machine object: %v",
-			t.string(),
-			err,
-		)
-		return fmt.Errorf("%s: failed to delete machine: %v", t.string(), err)
+	logger.Info("Applying unhealthy annotation")
+	if err := patchHelper.Patch(ctx, t.Machine); err != nil {
+		return fmt.Errorf("%s: failed to annotate machine: %v", t.string(), err)
 	}
-	r.Eventf(
-		t.Machine,
-		corev1.EventTypeNormal,
-		EventMachineDeleted,
-		"Machine %v has been remediated by requesting to delete Machine object",
-		t.string(),
-	)
```
Yeah I wasn't sure about these events, it wasn't super specified and seemed less useful now that the event is just annotation, but I don't have strong feelings about it.

benmoss commented:
apidiff is failing because of 

```
  Incompatible changes:
  - EventMachineDeleted: removed
  - EventMachineDeletionFailed: removed
  - EventSkippedControlPlane: removed
```

Do we have to leave these here? We should probably be triggering the first two from the MachineSet controller remediation, but the third won't be happening anymore.

vincepri commented on controllers/controllersfakes/fake_machine_remediator.go:
```diff
@@ -1,4 +1,19 @@
-// Code generated by counterfeiter. DO NOT EDIT.
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
```
In CAPA we have had some automation to get these files the correct headers

benmoss commented:
/hold adding some integration coverage

detiber commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -342,7 +342,8 @@ var _ = Describe("MachineHealthCheck Reconciler", func() {
 					key := types.NamespacedName{Namespace: m.Namespace, Name: m.Name}
-					Expect(k8sClient.Get(ctx, key, machine)).ToNot(Succeed())
+					Expect(k8sClient.Get(ctx, key, machine)).To(Succeed())
+					Expect(machine.GetAnnotations()).To(HaveKeyWithValue(clusterv1.MachineUnhealthy, ""))
```
```suggestion
					Expect(machine.GetAnnotations()).To(HaveKey(clusterv1.MachineUnhealthy))
```

JoelSpeed replied:
This was I change I had suggested (to check the value) since we are expecting it to be empty (based on the proposal), there had been talk of putting a value in this but that was decided against, I don't feel strongly about it though if you'd prefer not to check the value

detiber replied:
In the past we've removed explicit checks for empty string for keys where we only really care that the key is present. Unless we really care about the value for some reason I think we should stick with a key-based check.

detiber commented on util/healthcheck/machine.go:
```diff
@@ -0,0 +1,63 @@
+limitations under the License.
+*/
+
+package healthcheck
```
It seems odd to have this in a package called healthcheck, since it's not doing health check. Maybe rename the package `remediation` and then `UnhealthyMachineDeletionRemediator` could just be `UnhealthyMachineDeletion`?

vincepri commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -522,43 +518,22 @@ func TestClusterToMachineHealthCheck(t *testing.T) {
 	// BEGIN: Set up test environment
 	g := NewWithT(t)
 
-	testEnv = &envtest.Environment{
-		CRDs: []runtime.Object{
-			external.TestGenericBootstrapCRD,
-			external.TestGenericBootstrapTemplateCRD,
-			external.TestGenericInfrastructureCRD,
-			external.TestGenericInfrastructureTemplateCRD,
-		},
-		CRDDirectoryPaths: []string{filepath.Join("..", "config", "crd", "bases")},
-	}
```
Is this change required for the purpose of this PR?

benmoss replied:
Not strictly but I thought it was a good cleanup

vincepri replied:
Sounds good, in general I'd suggest to split changes in multiple steps, it's easier to look at changes if they're logically separated in multiple PRs

benmoss replied:
Yeah I'll pull it out into a separate PR

vincepri commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -522,43 +518,22 @@ func TestClusterToMachineHealthCheck(t *testing.T) {
-
-	var err error
-	cfg, err := testEnv.Start()
+	testEnv, err := helpers.NewTestEnvironment()
```
Looking in `NewTestEnvironment`, it seems we're trying to make a generic catch-all functions to create test environments, although the `CRDDirectoryPaths` will be different depending which package we're testing

benmoss replied:
I think we can just include all the CRDs all the time

vincepri commented on test/integration/mhc_test.go:
```diff
@@ -0,0 +1,253 @@
+
+func TestMachineSetsWithMachineHealthCheck(t *testing.T) {
+	g := NewWithT(t)
+	testEnv, err := helpers.NewTestEnvironment()
```
We've never tested creating more than one test environment, would this preclude us from running tests in parallel?

benmoss replied:
I think they're isolated environments if you use a new environment per-test, but we should also make our tests share environments more often and we can get some faster test runs that way. envtest supports `USE_EXISTING_CLUSTER` which will target an existing Kubernetes cluster.

benmoss commented:
> One thing that crossed my mind while reading this, should we have a check for when a Machine is already marked as unhealthy that we don't attempt to re-annotate it/we don't log that it's being marked for remediation again?
Done

> Secondly, are the fakes that are included in this PR necessary? For instance the util//healthcheck/machine_test.go tests could substitute the controller-runtime fake client and instead check that the machines were deleted rather than checking that delete was called the right number of times? I think it would make the tests here a lot easier to read and understand

1. controller-runtime fake client is [deprecated](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/fake/doc.go#L20-L21)
2. It's a smaller interface than the full client, it's better to rely on smaller interfaces where possible.

> Similarly when there's a fake remediator being passed into the MachineSet controller, I would question whether that's needed to be faked or whether you could just use the real remediator here? (I don't see any tests exercising the remediation within the MachineSet tests so that would also be good to add)

The purpose of using a mock here is so not to test the two objects together, but yeah the real remediator will just noop will all the existing tests since so I changed it to no longer use a mock.

benmoss commented:
Probably need to wait for #2995 to be merged first now

vincepri commented on controllers/healthcheck/machine.go:
```diff
@@ -0,0 +1,67 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//go:generate counterfeiter . Client
+type Client interface {
+	Delete(context.Context, runtime.Object, ...client.DeleteOption) error
+}
```
Could we use controller runtime fake client instead of creating a new mock client here?

benmoss replied:
Isn't it deprecated?

benmoss replied:
I responded to that here too https://github.com/kubernetes-sigs/cluster-api/pull/2975#issuecomment-622539961, I think it's better to use a narrower interface where possible, and we don't need the integration coverage of a fake object database that the CR client gives us

vincepri replied:
Yes, but we shouldn't worry about it until CR 1.0, and we have https://github.com/kubernetes-sigs/controller-runtime/issues/768 where community is discussing potential approaches

vincepri commented on controllers/healthcheck/machine.go:
```diff
@@ -0,0 +1,67 @@
+	logger logr.Logger
+}
+
+func (r *UnhealthyMachineDeletionRemediator) Remediate(ctx context.Context, machines []clusterv1.Machine) (bool, error) {
```
Should this satisfy an interface?

benmoss replied:
What do you mean?

vincepri replied:
For example,

```go
type Strategy interface {
    Remediate(...) ...
}
```

vincepri commented on controllers/healthcheck/machine.go:
```diff
@@ -0,0 +1,67 @@
+	}
+}
+
+type UnhealthyMachineDeletionRemediator struct {
```
```suggestion
type OwnerControllerStrategy struct {
```

If we have these in a package called `remediators` then in the future we could have multiple strategies in there

benmoss replied:
This is still the deletion strategy though, right? They will all be owner controller strategies in that we expect the owner controller to be the one running the remediator.

vincepri replied:
For now, yes that's correct. In the future if we have multiple strategies I can see us having some logic to pick a strategy based on what the MHC is configured to. One of them could be a no-op strategy, which falls in line with external remediation.

We can always rework it later, but if we structure the code to be composable today, it's a good investment for future iterations, that's also why I was suggesting to have a `remediation.Strategy` interface above.

vincepri replied:
The package can/should be under `controllers/healthcheck/remediation` potentially

vincepri commented on controllers/healthcheck/machine.go:
```diff
@@ -0,0 +1,67 @@
+		deleted = true
+	}
+
+	return deleted, kerrors.NewAggregate(errs)
```
Should we instead return the number of deleted Machines?

benmoss replied:
We could, but all we care about is whether we deleted any or none as far as I can think of

vincepri replied:
I can see that some clients might want to have the number of successfully deleted machines and compare it to the input

sedefsavas commented on test/helpers/envtest.go:
```diff
@@ -0,0 +1,141 @@
+	utilruntime.Must(apiextensionsv1.AddToScheme(scheme))
+	utilruntime.Must(clusterv1.AddToScheme(scheme))
+	utilruntime.Must(bootstrapv1.AddToScheme(scheme))
+
```
Should we add controlplane scheme too?

benmoss replied:
It can be added when someone writes a test using it, I only did what was necessary for the controllers package for now.

benmoss commented:
/hold cancel

benmoss commented:
envtest helper is still coming form #2995 so once that is merged it should disappear from the commits here

benmoss commented:
Do we want `EventMachineDeleted` / `EventMachineDeletionFailed` to be emitted by the MachineSet controller now?

vincepri commented:
/milestone v0.3.6

vincepri commented:
@benmoss Might be worthwhile to keep those events for now, wdyt?

sedefsavas commented on controllers/machinehealthcheck_targets_test.go:
```diff
@@ -337,68 +331,32 @@ func TestHealthCheckTargets(t *testing.T) {
 func TestRemediate(t *testing.T) {
```
TestAnnotate?

benmoss commented on controllers/machinehealthcheck_controller.go:
```diff
@@ -242,11 +242,11 @@ func (r *MachineHealthCheckReconciler) reconcile(ctx context.Context, cluster *c
 		logger.V(3).Info("Target meets unhealthy criteria, triggers remediation", "target", t.string())
-		if err := t.remediate(ctx, logger, r.Client, r.recorder); err != nil {
+		if err := t.annotate(ctx, logger, r.Client, r.recorder); err != nil {
 			logger.Error(err, "Error remediating target", "target", t.string())
```
Included this in my latest commit

benmoss commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -342,7 +342,8 @@ var _ = Describe("MachineHealthCheck Reconciler", func() {
 				for _, m := range rtc.expectRemediated() {
```
Also caught a bug where I accidentally left the assertion

```
Expect(machine.GetDeletionTimestamp().IsZero()).To(BeTrue())
```

on the `expectNotRemediated` machines

vincepri commented on test/integration/mhc_test.go:
```diff
@@ -0,0 +1,240 @@
+// +build integration
+
+/*
+Copyright 2019 The Kubernetes Authors.
```
```suggestion
Copyright 2020 The Kubernetes Authors.
```

benmoss commented:
/test pull-cluster-api-e2e

enxebre commented on controllers/machineset_controller.go:
```diff
@@ -211,6 +211,31 @@ func (r *MachineSetReconciler) reconcile(ctx context.Context, cluster *clusterv1
+		logger.Info("Failed while deleting unhealthy machines", "err", err)
+		return ctrl.Result{}, errors.Wrap(err, "failed to remediate machines")
+	} else if anyDeleted {
+		// Do not do any other operations while waiting for deletion to finish
```
can we clarify why do we need to "Do not do any other operations while waiting for deletion to finish"?

enxebre commented:
we should reflect in docs/release notes that any entity able to add annotations to the resource has effectively permissions to delete machines.

ncdc commented:
So with this change, someone who has `edit machines` permissions but not `delete` is able to circumvent that... Hmmm 🤔

vincepri commented:
/hold

benmoss commented:
Since this issue is so long already and we are going to be revising the strategy I am going to close this and will reopen a new PR when it's ready

vincepri commented:
/milestone v0.3.x
